### DISCUSSION TITLE
Update a blob field

### DISCUSSION THREAD

**Discussion by lbensch** [fluentmigrator/fluentmigrator#1340] [Q&A] [answered]
Hello, I would like to update a blob field with a text file, but I can't seem to find how to do this with fluentmigrator. My file is an embbeded resource, and when I try to do a set with this field, I get an error. Could you help me ?

**Comment by jzabroski** [comment]
@lbensch What error do you get?  What version of FLuentMigrator are you using?  What .net target framework? etc

**Comment by lbensch** [comment]
I use :
- .Net standard 2.0
- fluentmigrator 3.2.1

and my error :
{"Unable to complete network request to host \"\r\nNo message for error code 335544721 found."}

**Comment by lbensch** [comment]
It's not so much the mistake that I want to point out, I think I'm doing it wrong. I would just like to know how to update a blob.

**Reply by jzabroski** [reply]
Are you sure the issue is with you calling FluentMigrator incorrectly?  It looks like the error code you posted is for Firebird.  See: https://stackoverflow.com/a/35982783/1040437

Separately, I found a reply by cincura.net (Jiří Činčura, who is the maintainer of the .NET Firebird driver) on Stack Overflow, providing guidance on how to load blob data into .NET Firebird driver: https://stackoverflow.com/a/23150735/1040437

It looks like, at least as of 2014, there were some hard limitations on the size of the blob.  So are you using a 32 bit process?  And is that limitation still valid?  When I spoke to the maintainer last year, he indicated there is a Firebird mailing list for these types of questions.  He also pointed me here for filing any bugs or questions: https://github.com/cincuranet/FirebirdSql.Data.FirebirdClient/blob/master/Provider/readme.txt#L24

**Reply by jzabroski** [reply]
The problem might _also_ be that we still reference and older version of the Firebird driver in the FluentMigrator Runners.  I would first try seeing if you can call the Firebird driver directly with a console program, then figure out how to get it working in FluentMigrator.  If it is a bug, I am happy to do a urgent hotfix for the 3.2.8 code.

**Comment by lbensch** [comment]
I am using SQL Hammer to update my blobs directly, and it works fine with my file. If I understand what you are telling me, the problem would come from the .Net Firebird version? But how can I change this?

**Reply by jzabroski** [reply]
I have no idea what SQLHammer is.  I searched nuget.org and got 0 results.  https://www.nuget.org/packages?q=sqlhammer 

I found a [blog post from 2007](https://www.firebirdnews.org/blazetop-sqlhammer-17-a-tool-for-ibfb-database-administration/) indicating there was some tool for FIrebird named SQL Hammer that got named to BlazeTop - if this is true, I'm going to ask you be a lot more specific for help and not use obscure product references to products that no longer exist. 

I will repeat my recommendation to use the .NET Firebird Driver and C# console program to prove that you can write a C# program that can load data into your Firebird program.  Right now, you just aren't giving me enough information to help you.

**Reply by lbensch** [reply]
SQLHammer  is a tool, not a nuget.

**Comment by lbensch** [comment]
I changed the versions used:
- fluentmigrator 3.2.9
- firebirdclient 7.5.0

Using connection string User=SYSDBA;Password=********;Database=C:\Program Files (x86)\AxiSoftware\Database\AXI_DATABASE.FDB;DataSource=127.0.0.1;Port = 3055; Dialect = 3; Charset = NONE; Role =; Connection lifetime = 15; Pooling = true;MinPoolSize = 0; MaxPoolSize = 50; Packet Size = 16384; ServerType = 0;

But I still have the problem.

**Comment by lbensch** [comment]
I created a small console program in .netcore 3.1 with firebirdclient .net 7.5.0.

`
        static void Main()
        {
            using var conn = new FbConnection(@"User=SYSDBA;Password=masterkey;Database=C:\Program Files (x86)\AxiSoftware\Database\AXI_DATABASE.FDB;DataSource=127.0.0.1;Port = 3055; Dialect = 3; Charset = NONE; Role =; Connection lifetime = 15; Pooling = true;MinPoolSize = 0; MaxPoolSize = 50; Packet Size = 16384; ServerType = 0");
            conn.Open();
            Console.WriteLine("Connected!");

            var assembly = Assembly.GetExecutingAssembly();
            var resourceName = assembly.GetManifestResourceNames()
                .Single(str => str.EndsWith("AxiRB_0153_OrdoEPRNonAldA4Print.rtm"));
            Stream stream = assembly.GetManifestResourceStream(resourceName);

            StreamReader r = new StreamReader(stream ?? throw new InvalidOperationException());

            var content = r.ReadToEnd();
                
            string updateSQL =  @"UPDATE PRETAT SET PRE_STRUCTURE = @DATA WHERE PRETAT_PK = 153";
            using FbCommand command = new FbCommand( updateSQL, conn );
            try
            {
                byte[] buffer = new byte[stream.Length];
                stream.Read(buffer, 0, buffer.Length);

                command.Parameters.Add("@DATA", FbDbType.Binary).Value = Encoding.UTF8.GetBytes(content);//buffer;
                command.ExecuteNonQuery();
                command.Dispose();
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                throw;
            }
        }
`

Everything works well.

**Reply by jzabroski** [reply]
Thanks, that is very helpful.  I think I found the problem: https://stackoverflow.com/a/8558664/1040437

> You are adding text inline in the query. Then there are limits: first query text in Firebird is limited to 64 kB, there is also a limit to the varchar size.
> 
> If you want to add more data to a blob, you need to stream it to the blob using parametrized queries. See for an example: http://www.firebirdsql.org/en/net-examples-of-use/#update_text_blob_field

FluentMigrator is probably not streaming the blob, but instead pasting it as text.  I bet the the full sql text exceeds the 64 kB limit.

As a workaround, you can take your example program and call Execute.WithConnection using your above program.  Note that Execute.WithConnection is executed "out of sequence" from the flow of control of a migration, which is something that trips people up the first time they use it.

**Reply by jzabroski** [reply]
@lbensch Just so when I go to fix this I cover the right path through the system, is your migration from the embedded text file something like the following pseudo-code:

```c#
    [Migration(0, "This demonstrates how to reproduce FluentMigrator Discussion #1340.")]
    
    public class M0_DummyMigration : ForwardOnlyMigration
    {
        public override void Up()
        {
// Assume EmbeddedResourceUtility exists, and
// GetData gets the manifest resource stream and casts the stream to the generic type (here, byte[]).
            Update.Table("PRETAT")
                .Set(new { PRE_STRUCTURE = EmbeddedResourceUtility.GetData<byte[]>("YourLocalDataResourceName") })
                .Where(new { PRETAT_PK = 153 });
        }
    }
```

This would still not be perfect, because for really large streams, FluentMigrator needs a way to lazily evaluate the stream.  But the general idea is there.

**Comment by lbensch** [comment]
I am not sure I understand, is it a code that you recommend or is it an evolution to come ?

**Reply by jzabroski** [reply]
Probably an evolution to come.  I honestly don't know what we do, across each database provider, if you pass in a byte array.

**Comment by lbensch** [comment, marked as answer]
I do something like this : 
```c#
            var resourceName = assembly.GetManifestResourceNames().Single(str => str.EndsWith("AxiRB_0153_OrdoEPRNonAldA4Print.rtm"));
            using (Stream stream = assembly.GetManifestResourceStream(resourceName))
            {
                using (StreamReader r = new StreamReader(stream ?? throw new InvalidOperationException()))
                {
                    var content = r.ReadToEnd();

                    //Update.Table("PRETAT")
                    //    .Set(new { PRE_STRUCTURE = Encoding.UTF8.GetBytes(content) })
                    //    .Where(new { PRETAT_PK = 153 });

                    Execute.WithConnection((conn, tran) =>
                    {
                        string updateSQL =  @"UPDATE PRETAT SET PRE_STRUCTURE = @DATA WHERE PRETAT_PK = 153";
                        using FbCommand command = new FbCommand( updateSQL, (FbConnection)conn );
                        try
                        {
                            command.Transaction = (FbTransaction) tran;
                            command.Parameters.Add("@DATA", FbDbType.Binary).Value = Encoding.UTF8.GetBytes(content);//buffer;
                            command.ExecuteNonQuery();
                            command.Dispose();
                        }
                        catch (Exception e)
                        {
                            Console.WriteLine(e);
                            throw;
                        }
                    });
                }
            }
```

And it work.

**Reply by jzabroski** [reply]
That sounds about right, but I would call the StreamReader constructor that accepts an encoding

```c#
using (StreamReader r = new StreamReader(stream ?? throw new InvalidOperationException(), System.Text.Encoding.UTF8))
```

That would simplify the code a bit as it eliminates an unnecessary encoding conversion.
